src/groth16/aggregate/accumulator.rs

```diff
@@ -0,0 +1,103 @@
+use crate::bls::Engine;
+use ff::{Field, PrimeField};
+use groupy::{CurveAffine, CurveProjective};
+use paired::PairingCurveAffine;
+use rand::thread_rng;
+
+/// PairingCheck represents a check of the form e(A,B)e(C,D)... = T. Checks can
+/// be aggregated together using random linear combination. The efficiency comes
+/// from keeping the results from the miller loop output before proceding to a final
+/// exponentiation when verifying if all checks are verified.
+/// It is a tuple:
+/// - a miller loop result that is to be multiplied by other miller loop results
+/// before going into a final exponentiation result
+/// - a right side result which is already in the right subgroup Gt which is to
+/// be compared to the left side when "final_exponentiatiat"-ed
+pub struct PairingCheck<E: Engine>(E::Fqk, E::Fqk);
+
+impl<E> PairingCheck<E>
+where
+    E: Engine,
+{
+    pub fn new() -> PairingCheck<E> {
+        Self(E::Fqk::one(), E::Fqk::one())
+    }
+
+    pub fn new_invalid() -> PairingCheck<E> {
+        Self(E::Fqk::one(), E::Fqk::zero())
+    }
+
+    pub fn from_pair(result: E::Fqk, exp: E::Fqk) -> PairingCheck<E> {
+        Self(result, exp)
+    }
+
+    pub fn from_miller_one(result: E::Fqk) -> PairingCheck<E> {
+        Self(result, E::Fqk::one())
+    }
+
+    /// returns a pairing tuple that is scaled by a random element.
+    /// When aggregating pairing checks, this creates a random linear
+    /// combination of all checks so that it is secure. Specifically
+    /// we have e(A,B)e(C,D)... = out <=> e(g,h)^{ab + cd} = out
+    /// We rescale using a random element $r$ to give
+    /// e(rA,B)e(rC,D) ... = out^r <=>
+    /// e(A,B)^r e(C,D)^r = out^r <=> e(g,h)^{abr + cdr} = out^r
+    /// (e(g,h)^{ab + cd})^r = out^r
+    pub fn from_miller_inputs<'a, I>(it: I, out: &'a E::Fqk) -> PairingCheck<E>
+    where
+        I: IntoIterator<Item = &'a (&'a E::G1Affine, &'a E::G2Affine)>,
+    {
+        let mut rng = thread_rng();
```

Comment:
should probably be OSRng, for better security gurantees

Reply:
Done in the https://github.com/filecoin-project/bellperson/pull/166